### PR TITLE
externalresources(engine): implement GCCoordinator & GCRunner

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #5513

### What is changed and how it works?
- Implemented `GCCoordinator`.
- Implemented `GCRunner`.

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/pkg/externalresource/manager/gc_coordinator.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+	c.notifier.GCNotify()
+	return nil
+}
+
+func (c *DefaultGCCoordinator) gcByOfflineExecutorID(ctx context.Context, executorID model.ExecutorID) error {
+	log.L().Info("Cleaning up resources meta for offlined executor",
+		zap.String("executor-id", string(executorID)))
+	return c.metaClient.DeleteResourcesByExecutorID(ctx, string(executorID))
+}
```
Comment (amyangfei) [inline engine/pkg/externalresource/manager/gc_coordinator.go:239]: If we don't need to perform GC operation for offline executor, does this mean all resources are bounded to one executor, what if we support S3 storage as resource in the future.

File: engine/pkg/externalresource/manager/gc_runner.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+	handler, exists := r.gcHandlers[tp]
+	if !exists {
+		log.L().Warn("no gc handler is found for given resource type",
+			zap.Any("resource-id", res.ID))
+	}
+
+	if err := handler(ctx, res); err != nil {
+		return err
+	}
```
Comment (amyangfei) [inline engine/pkg/externalresource/manager/gc_runner.go:124]: If not exists, `handler` is nil.

Comment (liuzix) [reply engine/pkg/externalresource/manager/gc_runner.go]: Fixed

File: engine/pkg/externalresource/manager/gc_coordinator.go (hunk 1 above)
Comment (liuzix) [reply engine/pkg/externalresource/manager/gc_coordinator.go]: We will adjust this when we implement S3 support. The responsibility of checking whether to perform actual GC should belong to the gcHandler for the resource types.

File: engine/pkg/externalresource/manager/gc_coordinator.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+			continue
+		}
+
+		err = c.runGCEventLoop(ctx, jobReceiver.C, executorReceiver.C)
+		jobReceiver.Close()
+		executorReceiver.Close()
+
+		if gerrors.Is(err, context.Canceled) || gerrors.Is(err, context.DeadlineExceeded) {
+			return errors.Trace(err)
+		}
```
[the remark this answers is not in the document]
Comment (liuzix) [reply engine/pkg/externalresource/manager/gc_coordinator.go:87]: Added logging.

File: engine/pkg/externalresource/manager/gc_coordinator.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+		jobReceiver, executorReceiver, err := c.initializeGC(ctx)
+		if err != nil {
+			log.L().Warn("GC error", zap.Error(err))
+			rl.Take()
+			continue
+		}
+
+		err = c.runGCEventLoop(ctx, jobReceiver.C, executorReceiver.C)
```
Comment (maxshuang) [inline engine/pkg/externalresource/manager/gc_coordinator.go:81]: What condition will this loop run again?
I scan the whole pr and find runGCEventLoop will not return if no error happen.

File: engine/pkg/externalresource/manager/gc_coordinator.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	pkgOrm "github.com/pingcap/tiflow/engine/pkg/orm"
+)
+
+// DefaultGCCoordinator implements interface GCCoordinator.
+// It is responsible for triggering file resource garbage collection.
+type DefaultGCCoordinator struct {
+	executorInfos ExecutorInfoProvider
+	jobInfos      JobStatusProvider
+	metaClient    pkgOrm.ResourceClient
+	notifier      GCNotifier
```
Comment (maxshuang) [inline engine/pkg/externalresource/manager/gc_coordinator.go:39]: Does it designed to notify the executor to do some real gc through rpc?

File: engine/pkg/externalresource/manager/gc_coordinator.go (hunk 4 above)
Comment (liuzix) [reply engine/pkg/externalresource/manager/gc_coordinator.go:81]: You are right. This loop is for retrying on errors.

Comment (maxshuang) [reply engine/pkg/externalresource/manager/gc_coordinator.go:81]: So `Run` will be called multi-times?

Comment (liuzix) [reply engine/pkg/externalresource/manager/gc_coordinator.go:81]: No, it will not.

**Review by maxshuang — APPROVED** [review]

**Pull request merged**